### PR TITLE
[SPARK-4502][SQL] Parquet nested column pruning - foundation (2nd attempt)

### PR DESCRIPTION
**Pull request by ajacques**
(Link to Jira: https://issues.apache.org/jira/browse/SPARK-4502)

**This is a restart of apache/spark#21320. Most of the discussion has taken place over there. I've only taken it as a based to make changes based on the code review to help move things along.**

Due to the urgency of the upcoming 2.4 code freeze, I'm going to open this PR to collect any feedback. This can be closed if you prefer to continue to the work in the original PR.

### What changes were proposed in this pull request?
One of the hallmarks of a column-oriented data storage format is the ability to read data from a subset of columns, efficiently skipping reads from other columns. Spark has long had support for pruning unneeded top-level schema fields from the scan of a parquet file. For example, consider a table, contacts, backed by parquet with the following Spark SQL schema:

```
root
 |-- name: struct
 |    |-- first: string
 |    |-- last: string
 |-- address: string
```
Parquet stores this table's data in three physical columns: name.first, name.last and address. To answer the query

`select address from contacts`
Spark will read only from the address column of parquet data. However, to answer the query

`select name.first from contacts`
Spark will read name.first and name.last from parquet.

This PR modifies Spark SQL to support a finer-grain of schema pruning. With this patch, Spark reads only the name.first column to answer the previous query.

### Implementation
There are two main components of this patch. First, there is a ParquetSchemaPruning optimizer rule for gathering the required schema fields of a PhysicalOperation over a parquet file, constructing a new schema based on those required fields and rewriting the plan in terms of that pruned schema. The pruned schema fields are pushed down to the parquet requested read schema. ParquetSchemaPruning uses a new ProjectionOverSchema extractor for rewriting a catalyst expression in terms of a pruned schema.

Second, the ParquetRowConverter has been patched to ensure the ordinals of the parquet columns read are correct for the pruned schema. ParquetReadSupport has been patched to address a compatibility mismatch between Spark's built in vectorized reader and the parquet-mr library's reader.

### Limitation
Among the complex Spark SQL data types, this patch supports parquet column pruning of nested sequences of struct fields only.

### How was this patch tested?
Care has been taken to ensure correctness and prevent regressions. A more advanced version of this patch incorporating optimizations for rewriting queries involving aggregations and joins has been running on a production Spark cluster at VideoAmp for several years. In that time, one bug was found and fixed early on, and we added a regression test for that bug.

We forward-ported this patch to Spark master in June 2016 and have been running this patch against Spark 2.x branches on ad-hoc clusters since then.

### Known Issues
Highlighted in https://github.com/apache/spark/pull/21320#issuecomment-408271470

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
If there's no objection within few days, let me get this in cc @cloud-fan and @gatorsmile and make other works and comments separate.

@mallman, if we are all happy here, mind taking a look https://github.com/apache/spark/pull/21320#issuecomment-408271470 and https://github.com/apache/spark/pull/21320#issuecomment-406765851

I will fix my own comments as a follow by myself.

Will credit this to you FWIW.

**Comment by mallman** [comment]
> @mallman, if we are all happy here, mind taking a look https://github.com/apache/spark/pull/21320#issuecomment-408271470 and https://github.com/apache/spark/pull/21320#issuecomment-406765851

I'm working on the former. I don't understand what kind of "second PR" @gatorsmile is referring to in the latter comment. I know we need to do something to fix the ignored tests, but those tests are in this PR. Am I supposed to create a PR on this PR?

**Comment by HyukjinKwon** [comment]
> we need to do something to fix the ignored tests, but those tests are in this PR.

If this one got merged, yes, we could fix them in a separate PR to enable them. I remember there need some fixes in `ParquetReadSupport`? I believe we can all work in roughly parallel.

**Comment by gatorsmile** [comment]
Just FYI, we are unable to merge it if it has a correctness bug.

**Comment by HyukjinKwon** [comment]
@gatorsmile, just for clarification, you mean some regressions about correctness bug in existing features, right?

**Comment by gatorsmile** [comment]
I mean https://github.com/apache/spark/pull/21320#issuecomment-408271470

**Comment by HyukjinKwon** [comment]
Ah, I think that's a bug when this feature is enabled. How about this: since this PR is open already and that's a bug in the new feature, we can get this in, I address my comments and he works on that comment and fixes in `ParquetReadSupport`.

**Comment by mallman** [comment]
Hi @gatorsmile. Where do you see us at this point? Do you still want to get this into Spark 2.4?

**Comment by gatorsmile** [comment]
We are still targeting this to 2.4, but we need to fix all the identified bugs before merging it.

**Comment by gatorsmile** [comment]
Normally, we change the default to false or revert the whole PR if the bugs are found during the RC (release candidate) stage.

**Comment by HyukjinKwon** [comment]
Im that case, I hope my comments are addressed together before merging it in. They are non trivial and I dom't usually go with them if there isn't special reason for it.

**Comment by ajacques** [comment]
Where does that leave both of these PRs? Do we still want this one with the code refactoring or to go back to the original? Are there any comments for this PR that would block merging? I've set the default to false in this PR.

**Comment by gatorsmile** [comment]
If the revert is very risky, we normally change the default from true to false when the bug is identified in the RC stage.

**Comment by mallman** [comment]
> Where does that leave both of these PRs? Do we still want this one with the code refactoring or to go back to the original? Are there any comments for this PR that would block merging? I've set the default to false in this PR.

@ajacques Do you want to rebase your PR off my last commit? Then I think it's up to @HyukjinKwon and @gatorsmile on how to proceed.

**Comment by ajacques** [comment]
@mallman, sounds good I'll get this PR updated with your latest changes as soon as I can.

**Comment by mallman** [comment]
> This patch fails Scala style tests.

Hi @ajacques. I'm not sure if you're aware of this, but you can run the scalastyle checks locally with

```
sbt scalastyle
```

**Comment by ajacques** [comment]
These test failures are in Spark streaming. Is this just an intermittent test failure or actually caused by this PR?

**Comment by mallman** [comment]
> These test failures are in Spark streaming. Is this just an intermittent test failure or actually caused by this PR?

I was able to run the first failing test successfully. Can we get a retest, please?

**Comment by mallman** [comment]
> I was able to run the first failing test successfully. Can we get a retest, please?

@ajacques I just rebased and pushed my branch off of master. Perhaps the easiest thing to do would be for you to rebase again off of my branch and do another force-push to your branch.

**Comment by ajacques** [comment]
Anybody else able to reproduce this failure? It succeeded on my developer machine.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by mallman** [comment]
> Anybody else able to reproduce this failure? It succeeded on my developer machine.

It worked for me, too. Let's see what a retest does.

**Comment by mallman** [comment]
Oh dear. I don't know why we're getting all these sigkills. I think we're going to need another retest...

**Comment by mallman** [comment]
Success! Now where do we stand, @gatorsmile @HyukjinKwon ?

**Comment by ajacques** [comment]
Are there any other blockers to enabling this by default now that @mallman fixed the currently known broken queries?

**Comment by mallman** [comment]
> Are there any other blockers to enabling this by default now that @mallman fixed the currently known broken queries?

The functionality exercised by the ignored test in `ParquetSchemaPruningSuite.scala` is still broken. That's something we're hoping to fix in a follow on PR. This PR has to be merged first.

**Comment by ajacques** [comment]
@mallman: [This one](https://github.com/apache/spark/pull/21889/files#diff-0c6c7481232e9637b91c179f1005426aR120)? I just enabled it on my branch and the test passed. Was it fixed by your latest changes or am I missing something?

```
Expected:
struct<id:int,name:struct<middle:string>,address:string>

Actual:
fileSourceScanSchemata = {ArrayBuffer@12560} "ArrayBuffer" size = 1
 0 = {StructType@15492} "StructType" size = 3
  0 = {StructField@15494} "StructField(id,IntegerType,true)"
  1 = {StructField@15495} "StructField(name,StructType(StructField(middle,StringType,true)),true)"
  2 = {StructField@15496} "StructField(address,StringType,true)"
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+      Nil)
+  }
+
+  testSchemaPruning("select a single complex field and the partition column") {
+    val query = sql("select name.middle, p from contacts order by id")
+    checkScanSchemata(query, "struct<id:int,name:struct<middle:string>>")
+    checkAnswer(query, Row("X.", 1) :: Row("Y.", 1) :: Row(null, 2) :: Row(null, 2) :: Nil)
+  }
+
+  ignore("partial schema intersection - select missing subfield") {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala]: Just want to confirm it. This will not generate an incorrect result, right?

Comment (ajacques) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala]: @mallman can probably expand on this, but the logic is only enabled when the feature config is set to true, but it's not enabled by default. As written, this ignored test case actually passes.

Comment (mallman) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala]: This test is passing now. I think that I was so convinced that there was a problem in the Spark codebase—based on several confounding  factors—that I overlooked the fact that the failing test was actually alerting me to a problem in my patch. Lesson learned.

I am re-running the entire catalyst and sql-core test suite again locally with the default setting for schema pruning set to `true`. If those tests pass, I will publish a revised commit to my branch with the ignored test re-enabled and the new default value of the schema pruning configuration. I will ask @ajacques to rebase off of my branch at that time. If that goes well, then I suggest we merge this PR to master.

Incidentally, before closing SPARK-4502, we should consider how to document and track future work on this functionality in Jira. I see at least three more PRs following this one in order to get back to the functionality of #16578. First is to restore support for displaying the number of physical columns that will be read in the physical query plan. That is *invaluable* in verifying that schema pruning is working as expected. Second is to restore support for pruning query filters ("WHERE" clauses). Following that is support for pruning in joins and aggregations in one or two more PRs. I may be forgetting some things. That will cover a good deal of use cases for analytic queries. There are some additional use cases I have in my back pocket that can be worked on, too, such as support for window queries.

So... do we create sub-issues in SPARK-4502? Or do we close SPARK-4502 and create new issues?

Overall, after 2+ years in review, I'm excited to see us *this close* to merging something in. Once we have this in master, I believe it will be ready for an ORC expert to contribute support for schema pruning for that file format. Support for other columnar file formats should theoretically be possible, too.

**Comment by ajacques** [comment]
@mallman: I've rebased on top of your changes and pushed. I'm seeing the following:

Given the following schema:
```
root
 |-- id: integer (nullable = true)
 |-- name: struct (nullable = true)
 |    |-- first: string (nullable = true)
 |    |-- middle: string (nullable = true)
 |    |-- last: string (nullable = true)
 |-- address: string (nullable = true)
 |-- pets: integer (nullable = true)
 |-- friends: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- first: string (nullable = true)
 |    |    |-- middle: string (nullable = true)
 |    |    |-- last: string (nullable = true)
 |-- relatives: map (nullable = true)
 |    |-- key: string
 |    |-- value: struct (valueContainsNull = true)
 |    |    |-- first: string (nullable = true)
 |    |    |-- middle: string (nullable = true)
 |    |    |-- last: string (nullable = true)
 |-- p: integer (nullable = true)
```

The query: `select name.middle, address from temp` throws:
```
Caused by: org.apache.parquet.io.ParquetDecodingException: Can not read value at 0 in block -1 in file file:/private/var/folders/ss/cw601dzn59b2nygs8k1bs78x75lhr0/T/spark-cab140ca-cbba-4dc1-9fe5-6ae739dab70a/contacts/p=2/part-00000-91d2abf5-625f-4080-b34c-e373b89c9895-c000.snappy.parquet
  at org.apache.parquet.hadoop.InternalParquetRecordReader.nextKeyValue(InternalParquetRecordReader.java:251)
  at org.apache.parquet.hadoop.ParquetRecordReader.nextKeyValue(ParquetRecordReader.java:207)
  at org.apache.spark.sql.execution.datasources.RecordReaderIterator.hasNext(RecordReaderIterator.scala:39)
  at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
  at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:109)
  at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:186)
  ... 20 more
Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
  at java.util.ArrayList.rangeCheck(ArrayList.java:657)
  at java.util.ArrayList.get(ArrayList.java:433)
  at org.apache.parquet.io.GroupColumnIO.getFirst(GroupColumnIO.java:99)
  at org.apache.parquet.io.GroupColumnIO.getFirst(GroupColumnIO.java:99)
  at org.apache.parquet.io.PrimitiveColumnIO.getFirst(PrimitiveColumnIO.java:97)
  at org.apache.parquet.io.PrimitiveColumnIO.isFirst(PrimitiveColumnIO.java:92)
  at org.apache.parquet.io.RecordReaderImplementation.<init>(RecordReaderImplementation.java:278)
  at org.apache.parquet.io.MessageColumnIO$1.visit(MessageColumnIO.java:147)
  at org.apache.parquet.io.MessageColumnIO$1.visit(MessageColumnIO.java:109)
  at org.apache.parquet.filter2.compat.FilterCompat$NoOpFilter.accept(FilterCompat.java:165)
  at org.apache.parquet.io.MessageColumnIO.getRecordReader(MessageColumnIO.java:109)
  at org.apache.parquet.hadoop.InternalParquetRecordReader.checkRead(InternalParquetRecordReader.java:137)
  at org.apache.parquet.hadoop.InternalParquetRecordReader.nextKeyValue(InternalParquetRecordReader.java:222)
  ... 25 more
```

No root cause yet, but I noticed this while working with the unit tests.

**Comment by mallman** [comment]
> @mallman: I've rebased on top of your changes and pushed. I'm seeing the following

That's the test case that I "unignored". It was passing. There must be some simple explanation. I will look into it.

**Comment by mallman** [comment]
> @mallman: I've rebased on top of your changes and pushed. I'm seeing the following:

That test passes for me locally. Also, I inspected your branch and could not find any errors in the rebase. What commit hash are you testing locally? I'm using `92901da3785ce94db501a4c3d9be6316cfbf29a9`.

Please ensure we're on the same commit. If so, try doing an `sbt clean` and running your test again.

**Comment by ajacques** [comment]
The tests as committed pass for me, but I removed the `order by id` and I got that error. Are you saying it works with the specific query in my comment?

**Comment by mallman** [comment]
> The tests as committed pass for me, but I removed the order by id and I got that error. Are you saying it works with the specific query in my comment?

Oh! I didn't notice you changed the query.

Okay. I'll take a closer look.

**Comment by mallman** [comment]
> The tests as committed pass for me, but I removed the order by id and I got that error. Are you saying it works with the specific query in my comment?

@ajacques Please try this query:

```
select id, name.middle, address from temp
```

**Comment by mallman** [comment]
> Test build #94228 has finished for PR 21889 at commit 92901da.

The test failure appears to be unrelated to this PR.

Is it just me or has the test suite become flakier in the past few months?

**Comment by mallman** [comment]
> select id, name.middle, address from temp - Works
> select name.middle, address from temp - Fails
> select name.middle from temp - Works
> select name.middle, id, address from temp - Works
> select name.middle, address, id from temp - Works

Removing the `order by` clause from your test query caused it to fail, but it has nothing to do with ordering. It appears that the failure in this case is manifested when the file scan schema is exactly the `name.middle` and `address` columns. Introducing the `order by` clauses in the test suite queries gave them necessary determinism for checking query answers, but these modifications also altered the file scan schema.

I need to fix the tests, but I think that the failure underlying the previously ignored test case has not been resolved after all. It was just a case of confusing coincidence. Unfortunately we're still not ready to merge this PR yet.

**Comment by mallman** [comment]
I've pushed a commit to restore the original test coverage while also ensuring determinism of the output. Don't ask me how I did it. It's a secret!

The test that was failing before it was kinda passing is now failing again so I marked it ignored so it wouldn't break Jenkins. And I reverted the commit that enabled this feature by default, because it's still broken.

**Comment by ajacques** [comment]
@mallman Is it related to [this revert in ParquetReadSupport](https://github.com/apache/spark/pull/21889/commits/0312a5188f0d6c9fc5304195dbdc703bf0aa3fb7#diff-245e70c1f41e353e34cf29bd00fd9029L86). I re-added this logic and all 32 tests in ParquetSchemaPruningSuite passed.

**Comment by mallman** [comment]
> @mallman Is it related to this revert in ParquetReadSupport.scala? I re-added this logic and all 32 tests in ParquetSchemaPruningSuite passed.

Yes. That's what we need to work on in the next PR.

**Comment by ajacques** [comment]
Alright to make sure we're all on the same page, it sounds like we're ready to merge this PR pending:
* Successful build by Jenkins
* Any PR comments from a maintainer

This feature will be merged in disabled state and can't be enabled until the next PR is merged, but we do not expect any regression in behavior in the default disabled state.

**Comment by mallman** [comment]
> Alright to make sure we're all on the same page, it sounds like we're ready to merge this PR pending:
>
> * Successful build by Jenkins
> * Any PR comments from a maintainer
>
> This feature will be merged in disabled state and can't be enabled until the next PR is merged, but we do not expect any regression in behavior in the default disabled state.

I agree.

**Comment by ajacques** [comment]
Jenkins build successful. Any PR comments/blockers to merge for phase 1?

cc @HyukjinKwon, @gatorsmile, @cloud-fan

**Comment by ajacques** [comment]
Is there anything I can do to help with this PR?

**Comment by HyukjinKwon** [comment]
Can we address the comments I left on that PR too? Looks that's the only way to get through this? FWIW, since https://github.com/apache/spark/commit/51bee7aca13451167fa3e701fcd60f023eae5e61 is merged, we can now contribute to all people involved here.

**Comment by HyukjinKwon** [comment]
Assuming from https://github.com/apache/spark/pull/21889#issuecomment-408330791, we shouldn't have any identified bug here. What kind of bugs left to be fixed?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala (hunk 1 above)
Comment (mallman) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala]: Sorry I did not follow up on this thread. The apparent test passing was actually caused by a change in the tests to sort the query results before comparison. This introduced a subtle side-effect that masked the failure in the ignored test. I eliminated that side-effect and found that the test was failing again. The test is now marked as ignored again.

The test failure manifests itself as an exception thrown during read. I've reset the default value of the schema pruning configuration flag to `false` to ensure end users do not encounter this problem should it not be corrected before release.

**Comment by mallman** [comment]
> Assuming from #21889 (comment), we shouldn't have any identified bug here. What kind of bugs left to be fixed?

That bug was address by b50ddb4. We still need to fix the bug underlying the failing (ignored) test case. I have a tentative fix for that, but @gatorsmile wants to review it in a follow-on PR.

**Comment by HyukjinKwon** [comment]
> but @gatorsmile wants to review it in a follow-on PR.

Where did he say it after the comment above?

Also why don't you address my comments if you're going to push more changes then.

**Comment by ajacques** [comment]
@HyukjinKwon Looks like most of your comments have been already addressed, but I've gone ahead and made a few more tweaks to help this get merged. Please let me know if any blocking comments have been missed.

As mentioned: This feature is not known to have any regressions in the default, disabled state.

**Comment by HyukjinKwon** [comment]
> but @gatorsmile wants to review it in a follow-on PR.

I need a confirmation from @gatorsmile. I don't want to ignore his decision here in

> Just FYI, we are unable to merge it if it has a correctness bug.

@ajacques, thanks. I overlooked the recent changes made. Will take another look soon but don't block on this since most of them look addressed from a cursory look.

**Comment by ajacques** [comment]
>> but @gatorsmile wants to review it in a follow-on PR.

> Where did he say it after the comment above?

It was my interpretation of this comment: https://github.com/apache/spark/pull/21320#issuecomment-406353694

@gatorsmile, @HyukjinKwon Do we wish to block this PR to fix the issue with it enabled? It's not clear what your expectations are for this PR. 
1. Are you okay with it not 100% working if it's disabled by default
2. Do you want this issue to be fixed at the cost of bringing more changes into this PR?

**Comment by HyukjinKwon** [comment]
That comment is before https://github.com/apache/spark/pull/21889#issuecomment-408330791. I am okay in general but want to be clear if I'm ignoring his decision or not.

**Comment by mallman** [comment]
See https://github.com/apache/spark/pull/21320#issuecomment-406353694 for @gatorsmile's request to move the changes to `ParquetReadSupport.scala` to another PR.

There was another, unrelated bug reported by @jainaks and addressed in https://github.com/apache/spark/pull/21320#issuecomment-408588685. AFAIK, there's nothing outstanding blocking this PR from being merged as I stated in 
https://github.com/apache/spark/pull/21889#issuecomment-410557228.

**Comment by HyukjinKwon** [comment]
just for clarification, so now .. there no outstanding bugs, some tests are ignored per https://github.com/apache/spark/pull/21320#issuecomment-406353694 and left comments were mostly addressed. Did i understand correctly?

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by mallman** [comment]
> just for clarification, so now .. there no outstanding bugs, some tests are ignored per #21320 (comment) and left comments were mostly addressed. Did i understand correctly?

The ignored tests—and the scenarios they are intended to test—will fail with a runtime exception if this feature is enabled. I put forward a fix in `ParquetReadSupport.scala`, but @gatorsmile didn't want to address that in this PR. Otherwise, there are no known bugs with this patch.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by mallman** [comment]
Are we waiting for @gatorsmile's go-ahead and merge?

**Comment by ajacques** [comment]
@mallman, while we wait for the go-no-go, do you have the changes for the next PR ready? Is there anything you need help with?

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by ajacques** [comment]
@gatorsmile Do you think there is a on deterministic failure in this change that causes it to inconsistently fail?

**Comment by gatorsmile** [comment]
retest this please

**Comment by mallman** [comment]
>> @mallman, while we wait for the go-no-go, do you have the changes for the next PR ready? Is there anything you need help with?

> I have the hack I used originally, but I haven't tried finding a better solution yet. It could take some time to understand the underlying problem/incompatibility/misunderstanding/etc.

I spent some time yesterday digging deeper into why the hack I wrote worked, and I think I understand now. Practically speaking, my follow-on PR will be about the same as the commit I removed. However, I can support it with some explanatory comments instead of just "this throws an exception sometimes".

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ProjectionOverSchema.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * class is motivated by columnar nested schema pruning.
+ */
+private[execution] case class ProjectionOverSchema(schema: StructType) {
+  private val fieldNames = schema.fieldNames.toSet
+
+  def unapply(expr: Expression): Option[Expression] = getProjection(expr)
+
+  private def getProjection(expr: Expression): Option[Expression] =
+    expr match {
+      case a @ AttributeReference(name, _, _, _) if fieldNames.contains(name) =>
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ProjectionOverSchema.scala]: Can we really avoid `AttributeReference(name, _, _, _)` pattern per https://github.com/databricks/scala-style-guide#pattern-matching?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ProjectionOverSchema.scala [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        Some(a.copy(dataType = schema(name).dataType)(a.exprId, a.qualifier))
+      case GetArrayItem(child, arrayItemOrdinal) =>
+        getProjection(child).map { projection => GetArrayItem(projection, arrayItemOrdinal) }
+      case GetArrayStructFields(child, StructField(name, _, _, _), _, _, containsNull) =>
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ProjectionOverSchema.scala]: For instance, these `, _, _, _), _, _,` looks excessive.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SelectedFieldSuite.scala [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+       StructField("field1", IntegerType, nullable = false) ::
+       StructField("field2", StructType(
+         StructField("subfield1", IntegerType, nullable = false) :: Nil)) ::
+       StructField("field3", ArrayType(StructType(
+         StructField("subfield1", IntegerType, nullable = false) :: Nil))) :: Nil))) ::
+     StructField("col8", ArrayType(StructType(
+       StructField("field1", ArrayType(IntegerType, containsNull = false), nullable = false)
+         :: Nil))) :: Nil)
+
+  private val testRelation = LocalRelation(schema.toAttributes)
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SelectedFieldSuite.scala]: Shall we have multiple relations with split schema for each? It wouldn't necessarily have one big deep nested schema which makes hard to read.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+  private val briefContacts =
+    BriefContact(2, Name("Janet", "Jones"), "567 Maple Drive") ::
+    BriefContact(3, Name("Jim", "Jones"), "6242 Ash Street") :: Nil
+
+  case class ContactWithDataPartitionColumn(
+    id: Int,
+    name: FullName,
+    address: String,
+    pets: Int,
+    friends: Array[FullName] = Array(),
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala]: nit: `Array.empty[FullName]`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    relatives: Map[String, FullName] = Map(),
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala]: nit: `Map.empty[String, FullName]`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+      spark.read.parquet(path + "/contacts").createOrReplaceTempView("contacts")
+
+      testThunk
+    }
+  }
+
+  private val schemaEquality = new Equality[StructType] {
+    override def areEqual(a: StructType, b: Any): Boolean =
+      b match {
+        case otherType: StructType => a sameType otherType
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala]: nit: avoid infix notation per https://github.com/databricks/scala-style-guide#infix

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetTest.scala [hunk 8]
```diff
@@ -59,42 +59,68 @@ private[sql] trait ParquetTest extends SQLTestUtils {
   }
 
   /**
-   * Writes `data` to a Parquet file, which is then passed to `f` and will be deleted after `f`
+   * Writes `df` to a Parquet file, which is then passed to `f` and will be deleted after `f`
    * returns.
    */
-  protected def withParquetFile[T <: Product: ClassTag: TypeTag]
-      (data: Seq[T])
+  protected def asParquetFile
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetTest.scala]: I would avoid unrelated refactoring and revert this back.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruning.scala [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+      case op @ PhysicalOperation(projects, filters,
+          l @ LogicalRelation(hadoopFsRelation @ HadoopFsRelation(_, _,
+            dataSchema, _, _: ParquetFileFormat, _), _, _, _)) =>
+        val projectionRootFields = projects.flatMap(getRootFields)
+        val filterRootFields = filters.flatMap(getRootFields)
+        val requestedRootFields = (projectionRootFields ++ filterRootFields).distinct
+
+        // If requestedRootFields includes a nested field, continue. Otherwise,
+        // return op
+        if (requestedRootFields.exists { case RootField(_, derivedFromAtt) => !derivedFromAtt }) {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruning.scala]: Are we really unable to make some private functions and split some logics at least?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SelectedField.scala [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+/**
+ * A Scala extractor that builds a [[org.apache.spark.sql.types.StructField]] from a Catalyst
+ * complex type extractor. For example, consider a relation with the following schema:
+ *
+ *   {{{
+ *   root
+ *    |-- name: struct (nullable = true)
+ *    |    |-- first: string (nullable = true)
+ *    |    |-- last: string (nullable = true)
+ *    }}}
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SelectedField.scala]: indentation

**Comment by mallman** [comment]
@ajacques I added a commit to enable schema pruning by default. It's a little more complete than your commit to do the same. Please rebase off my branch and remove your commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruning.scala [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+          } else {
+            op
+          }
+        } else {
+          op
+        }
+    }
+
+  private def buildPrunedDataSchema(fileDataSchema: StructType,
+                                    requestedRootFields: Seq[RootField]) = {
```
[the remark this answers is not in the document]
Comment (ajacques) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruning.scala]: Will fix all this. Seems like this is an opportunity to catch some of this with static analysis to avoid wasting the reviewer's time. Is Scalastyle not configured to catch this stuff?

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruning.scala]: As far as I know, no. I use IntelliJ (integration with scalastyle) to catch basic style nits but rather then that, I do it manually. It took me a while to get used to it before FWIW.

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala (hunk 5 above)
Comment (ajacques) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala]: This is a definition in a constructor, so I don't think we can do:
```scala
case class Contact(
  [...]
  friends: Array.empty[FullName],
  relatives: Map.empty[String, FullName]
)
```
Scala wants a colon, so I opted for:
```scala
case class Contact(
  [...]
  friends: Array[FullName] = Array.empty,
  relatives: Map[String, FullName] = Map.empty
)
```

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by mallman** [comment]
> Due to the urgency of the upcoming 2.4 code freeze, I'm going to open this PR to collect any feedback. This can be closed if you prefer to continue to the work in the original PR.

That would be my preference, yes, especially if it means less work for you.

**Comment by HyukjinKwon** [comment]
>  I've only taken it as a based to make stylistic changes based on the code review to help move things along.

This PR doesn't only include stylistic changes. Since stylistic changes do not usually block a PR, mind fixing the PR description?

**Comment by mallman** [comment]
Essentially, this PR was created to take the management of #21320 out of my hands, with a view towards facilitating its incorporation into Spark 2.4. It was my suggestion, one based in frustration. In hindsight, I no longer believe this strategy is the best—or most expedient—approach towards progress. Indeed, I believe the direction of this PR has become orthogonal to its motivating goal, becoming a dispute between myself and @HyukjinKwon rather than a means to move things along.

I believe I can shepherd #21320 in a way that will promote greater progress. @ajacques, I mean no disrespect, and I thank you for volunteering your time, patience and effort for the sake of all that are interested in seeing this patch become a part of Spark. And I apologize for letting you down, letting everyone down. In my conduct leading up to the creation of this PR I did not act with the greatest maturity or patience. And I did not act in the best interests of the community.

No one has spent more time or more effort, taken more responsibility or exhibited more patience with this 2+ year patch-set-in-the-making than myself. I respectfully submit it is mine to present and manage, and no one else's. Insofar as I have expressed otherwise in the past, I admit my error—one made in frustration—and recant in hindsight.

@ajacques, at this point I respectfully assert that managing the patch set I submitted in #21320 is not your responsibility, nor is it anyone else's but mine. I ask you to close this PR so that we can resume the review in #21320. As I stated there, you are welcome to open a PR on https://github.com/VideoAmp/spark-public/tree/spark-4502-parquet_column_pruning-foundation to submit the changes you've made for review.

Thank you.

**Comment by ajacques** [comment]
Thanks for the response all. @mailman If it's really your preference, I will create a PR against that branch and close this one. My intention was never to take away from your efforts, and I still consider my work here to be just minor stylistic tweaks on top of your work. I did this as service to help bridge the divide and hopefully alleviate frustrations. But this has been a bit frustrating being stuck between two sides of this and changing merge strategies often and don't wish to continue being in between like this. 

As such, I will create a PR, but hope it does not dragged out to settle any differences in opinions between maintainers and submitters. My goal is to make sure this valuable feature gets merged so many can benefit.
